src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,296 @@
+                    case IfStatementSyntax ifStatement:
+                        ExpressionSyntax left, right;
+                        switch (ifStatement)
+                        {
+                            case { Condition: BinaryExpressionSyntax { OperatorToken.RawKind: (int)SyntaxKind.EqualsEqualsToken } binary }
+                                // Only suggest the fix on built-in `==` operators where we know we won't change behavior
+                                when semanticModel.GetSymbolInfo(binary).Symbol is IMethodSymbol { MethodKind: MethodKind.BuiltinOperator }:
+
+                                left = binary.Left;
+                                right = binary.Right;
+                                break;
+                            case { Condition: IsPatternExpressionSyntax { Expression: var patternInput, Pattern: ConstantPatternSyntax { Expression: var patternExpression } } }:
+                                left = patternInput;
+                                right = patternExpression;
+                                break;
+                            case { Condition: InvocationExpressionSyntax { Expression: var receiver, ArgumentList.Arguments: { Count: 2 } arguments } }
+                                when referenceEqualsMethod != null && referenceEqualsMethod.Equals(semanticModel.GetSymbolInfo(receiver, cancellationToken).Symbol):
+
+                                left = arguments[0].Expression;
+                                right = arguments[1].Expression;
+                                break;
+
+                            default:
+                                return null;
+                        }
+
+                        var parameterInBinary = left.IsKind(SyntaxKind.NullLiteralExpression) ? TryGetParameter(right)
+                            : right.IsKind(SyntaxKind.NullLiteralExpression) ? TryGetParameter(left)
+                            : null;
+                        if (parameterInBinary is null)
+                        {
+                            return null;
+                        }
+
+                        var throwStatement = ifStatement.Statement switch
+                        {
+                            ThrowStatementSyntax @throw => @throw,
+                            BlockSyntax { Statements: { Count: 1 } statements } => statements[0] as ThrowStatementSyntax,
+                            _ => null
+                        };
+
+                        if (throwStatement?.Expression is not ObjectCreationExpressionSyntax thrownInIf
+                            || !IsConstructorApplicable(thrownInIf, parameterInBinary))
+                        {
+                            return null;
+                        }
+
+                        return parameterInBinary;
+
+                    // this.field = param ?? throw new ArgumentNullException(nameof(param));
+                    case ExpressionStatementSyntax
```

Comment:
similarly, can you extract this case out to a helper?

Reply:
I found that since much of the complexity was in the pattern itself, the extracted logic didn't feel sufficiently complex to justify a single-use helper.